src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -332,15 +358,74 @@ public void CopyChangedValues(TStructuralType original)
                 }
                 else
                 {
-                    // Recursively patch the subtree.
-                    bool isDeltaType = TypedDelta.IsDeltaOfT(deltaNestedResource.GetType());
-                    Contract.Assert(isDeltaType, nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+                    // Recursively patch the subtree.                    
+                    Contract.Assert(TypedDelta.IsDeltaOfT(((object)deltaNestedResource).GetType()), nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+
+                    Type newType = deltaNestedResource.StructuredType;
+                    Type originalType = originalNestedResource.GetType();
+
+                    if (deltaNestedResource.IsComplexType && newType != originalType)
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, newType, originalType, deltaNestedResource.ExpectedClrType);
```

Comment:
>originalNestedResource [](http://example.com/codeflow?start=24&length=22)

Does returning null here (in the case that we don't find a common base type) change the current behavior of just keeping the original originalNestedResource?

Reply:
In iteration 3 you returned originalNestedResource if no common base type was found. In iteration 4 the behavior changed to return null if no common base type was found. Was this intentional?